url026791d51b_dhleong_SeriesTracker_v6_tgz-pJ8-stv6_SeriesTrackerJ8/src/stv6/handlers/util/CoverHandler.java

```diff
@@ -0,0 +1,181 @@
+			return name.matches("^[cC][oO][vV][eE][rR]\\..*$");
+		}
+		
+	};
+	
+	@Override
+	public boolean handle(Request request, Response response) {
+		// get the series 
+		BasicSeries series = null;
+		try {
+			String id_ = request.path.substring(
+					request.path.lastIndexOf('/')+1);
+			int id = Integer.parseInt(id_);
+			
+
+			Series s = Profile.getInstance()
+				.getSeries().getById(id);
+			if (!s.isManaged()) {
+				System.err.println("unmanaged series");
+				return showDefault(response);
+			}
+			
+			series = (BasicSeries) s;
+		} catch (StringIndexOutOfBoundsException e) {
+			// no valid id
+			return showDefault(response);
+		} catch (NumberFormatException e) {
+			// not valid id
+			return showDefault(response);
+		} catch (ClassCastException e) {
+			// untracked, no local path
+			return showDefault(response);
+		}
+		
+//		if (series == null) {
+//			return showDefault(response);
+//		}
+		
+		String localPath = series.getLocalPath();
+		File localFile = new File(localPath);
+		
+		File[] files = localFile.listFiles(FILENAME_FILTER);
+		if (files == null || files.length == 0) {
+			return showDefault(response);
+		}
+		
+		File theFile = files[0]; // pick first candidate
+		if (!theFile.exists()) {		
+//			System.err.println("Cannot find: " + path + " (" +
+//			        theFile.getAbsolutePath() + ")");
+			System.err.println("No such file: " + theFile.getAbsolutePath());
+			return showDefault(response);
+		}
+		if (!theFile.canRead()) {
+//			System.err.println("Cannot read: " + theFile.getAbsolutePath());
+			return showDefault(response);
+		}
+		
+		return showFile(response, theFile, request);
+	}
+
+	private boolean showDefault(Response response) {
+	    File def = new File("default-cover.png");
+	    if (def.exists())
+	        return showFile(response, def, null);
+	    
+	    return false;
+	}
+	
+	/**
+	 * 
+	 * @param response
+	 * @param theFile
+	 * @param request Basically only pass if you're interested in letting
+	 *     clients cache the thing. If you're not (IE: the "default" cover),
+	 *     just pass null
+	 * @return
+	 */
+	private boolean showFile(Response response, File theFile, Request request) {
+	    if (request != null && FileHandler.handleCache(
+	            request, response, theFile)) {
+	        return true;
+	    }
+	    
+		String name = theFile.getName();
+		String ext = name.substring(name.lastIndexOf('.')+1);
+		response.setContentType("image/"+ext);
+		byte[] imgdata = null;
+		try {
+			imgdata = getBytesFromFile(theFile);
+		} catch (IOException e) {
+			e.printStackTrace();
+			return showDefault(response);
+		}
+			
+		if (imgdata == null)
+			return showDefault(response);
+		
+		response.setBody(imgdata);
+		return true;
+	}
+
+	/** 
+	 * Thanks to laziness and Java Developer's Almanac
+	 * @param file
+	 * @return
+	 * @throws IOException
+	 */
+    public static byte[] getBytesFromFile(File file) throws IOException {
+        InputStream is = new FileInputStream(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *is*. The resource is referenced at line 162. The resource is closed at line 172. There are other execution paths that do not contain closure statements, for example, when an if condition at line 149 evaluates to true. Ensure *is* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).